internal/dataplane/kong_client.go

```diff
@@ -117,6 +117,11 @@ type KongClient struct {
 	// SHAs is a slice is configuration hashes send in last batch send.
 	SHAs []string
 
+	// lastValidKongState contains the last valid configuration pushed
+	// to the gateways. It is used as a fallback in case a newer config version is
+	// somehow broken.
+	lastValidKongState *kongstate.KongState
```

Comment:
For a moment I thought that we should correlate that with a hash of sorts so that users would get it an could correlate it with the prior config and know which particular config from the past they have configured but it occurred to me that SHAs that we have in the field above are created specifically for clients based on kong state and their e.g. plugin schema so we can't use that for **all** clients. 🤔 

Not sure if necessary but could be worth considering.

@mheap Would users be interested in this bit of information or is having the last known config is good enough without any metadata about it needed?

Reply:
Do we have a workflow in mind? We could add a hash field to the events, but the hash is narrowly intended for determining whether we need to push an update or not, and doesn't readily lend itself to other use cases:

- The hash provides a quick means of checking if two blobs are identical without comparing their contents line by line, but can't tell you what the actual differences are. You can't work backwards from the hash, essentially.
- The hashes aren't tied to a specific problem, and problems may persist across multiple hashes. If, for example, you add a misconfigured route in one iteration, don't fix it, and then update an endpoint/target on the next iteration, you'll have multiple different hashes that have the same problem.

We do have https://docs.konghq.com/kubernetes-ingress-controller/latest/troubleshooting/#dumping-generated-kong-configuration (and `GET /config` on the admin API) to show the current applied and desired config (identical if there are no issues) in a more human-usable format ("the config is literally this" versus "a random number").

We could maybe use the hashes to identify whether an event is currently relevant or not, but we need to think through the actual process of "alert on issue, direct user to remediate somehow, confirm problem clear" to ensure we're building what we need. Surfacing the value alone doesn't provide a complete troubleshooting workflow and path to solution.

Reply:
My intended workflow was "is this an error I've already acked or do I need to log it somewhere?" when using an event forwarder as events are emitted multiple times. Understood that you can't work backwards.

Reply:
In addition to storing the hash alongside the Kong state itself (I'll add it), to solve the issues @pmalek mentioned and the one I mentioned in the design doc (KIC restarts - Kong gets scaled up - new Kong pods have no configuration), we could use the already existing Kong pods as the source of truth for the last good configuration in case KIC loses the previous good configuration and restarts. If no last good configuration exists in KIC, but some Kong pods have a good configuration we can fetch it and store it as the last good one. WDYT?

Reply:
That's a great idea. Does it add much effort/maintenance complexity?

Reply:
I am not sure about the complexity, I'll investigate a bit and get back here with an estimation

Reply:
I'm wondering if this solution would be better than using a `ConfigMap`. The complexity seems to be similar (GET `/config` vs get `-o yaml` contents of a `ConfigMap`) yet we're forcing ourselves into a corner in situations were no Kong pods are available (e.g. there only 1 replica and it's that sole replica that crashed, got restarted or whatever else).

WDYT about this @mlavacca, @mheap? I'm curious what you think about the above and what pros and cons you can call out for each of those approaches.

Reply:
> The data stored in a ConfigMap cannot exceed 1 MiB

This isn't sufficient for many deployments

Reply:
I think I misread earlier. I see three possible scenarios:

* Proxy pods restart / scale out = KIC cache is the source of truth
* KIC restarts, broken config on API server = we fetch from existing proxy pod and cache
* Both KIC and proxy pods restart = fix your k8s API server. We can't help you

Reply:
Yeah, we cannot use configmaps because of the size limit (in the design docs I proposed using a PV for this reason).

> I think I misread earlier. I see three possible scenarios:

> - Proxy pods restart / scale out = KIC cache is the source of truth
> - KIC restarts, broken config on API server = we fetch from existing proxy pod and cache
> - Both KIC and proxy pods restart = fix your k8s API server. We can't help you

This is 100% what I meant with the previous comment.

I think the implementation choice boils down to how much is the last point acceptable: if we want to provide the highest level of robustness, then the only possible solution is a PV. Otherwise, if we agree upon having that failure case, I think this solution would be easier (and would not imply managing a PV). @mheap the case "Both KIC and proxy pods restart = fix your k8s API server. We can't help you" is acceptable from your point of view?